src/App.tsx

```diff
@@ -123,20 +127,27 @@ const persistentId = (key: string) => {
 }
 
 export const App = ({ actyx }: Props) => {
-  let effectHasRun = false
+
+  const plants = useRef<Map<string, Mgmt<PlantState>>>(new Map())
+  const robots = useRef<Map<string, Mgmt<RobotState>>>(new Map())
 
   const [plantId, setPlantId] = useState('')
   const [robotId, setRobotId] = useState('')
 
-  // all effects need to go in here, because useState setters are run many times
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, setRefreshSymbol] = useState(Symbol())
+
   useEffect(() => {
-    // React is weird
-    if (effectHasRun) return
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    else effectHasRun = true
+    let alive = true;
```

Comment:
why is this needed? shouldn’t it be enough to track cancellation callbacks and calling them?

Reply:
This is needed so that `runPlant` and `runRobot` know when the `useEffect's` lifetime ends.
The alternative to this will be a sync `runPlant` and `runRobot`, as you told me the last time, returning a cancellation callback which when called will flip a similar flag, although the `alive` flag will be inside each of `runPlant` and `runRobot`